core/src/main/kotlin/McuMgrResult.kt

```diff
@@ -0,0 +1,122 @@
+package com.juul.mcumgr
+
+import com.juul.mcumgr.McuMgrResult.Error
+import com.juul.mcumgr.McuMgrResult.Failure
+import com.juul.mcumgr.McuMgrResult.Success
+import com.juul.mcumgr.message.Response
+
+sealed class McuMgrResult<out T> {
```

Comment:
This result class doesn't feel domain-specific. It looks like just a general result class.
That's not necessarily a problem, but it may be unnecessarily adding a lot of complexity as opposed to just propagating exceptions.

If we're just wrapping exceptions for the purpose of wrapping in a general result object, consumers of the library could do that with `runCatching`. By wrapping in a result you've added overhead and consumers have to unwrap if they want the exception. Seems more optimal to give the choice of creating the wrapped objects?

Now if this were more domain-specific (i.e. some of the sealed sub classes represented result states, such as upload failed due to timeout, or specific responses from the remote device, then suddenly this result is very informative/valuable).

Reply:
Good point. I feel somewhat uncomfortable making public APIs so throw heavy, but it does provide a broader API for consumers to use.

Domain specific results might be better for wrapping classes (e.g. Downloader, Uploader) which perform additional logic on top of the base mcumgr command.

I'll keep McuMgrResult around to use for now, but I wont use it in the public API. If at the end,  McuMgrResult is no longer useful, I can remove it.

Reply:
> I feel somewhat uncomfortable making public APIs so throw heavy

Agreed, and I definitely don't want to discourage use of result objects as they are more idiomatic (vs. throwing exceptions).

If possible, it would be preferred to return a domain-specific result object.

In this case, the result object just wasn't domain-specific. It looked like part of the design challenge is that this result object was wrapping your transport functions; which themselves don't provide a well defined possible error states (so we lean on the very general throw mechanics). Perhaps clearly defined failure states could help form a domain-specific result object for your transport functions. Then at the mcumgr layer (that calls the transport functions) you could easily map those results, or basically handle the ones than can be handled and then return the mcumgr-specific result object?